pkg/destroy/gcp/network.go

```diff
@@ -0,0 +1,647 @@
+		routes, err := o.listNetworkRoutes(network.url)
+		if err != nil {
+			return err
+		}
+		for _, route := range routes {
+			if err = o.deleteRoute(route); err != nil {
+				return err
+			}
+		}
+
+		err = o.deleteNetwork(network.name)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (o *ClusterUninstaller) listNetworkRoutes(networkURL string) ([]string, error) {
+	return o.listRoutes(fmt.Sprintf("network=%q", networkURL))
+}
+
+func (o *ClusterUninstaller) listClusterRoutes() ([]string, error) {
+	return o.listRoutes(o.clusterIDFilter())
+}
+
+func (o *ClusterUninstaller) listRoutes(filter string) ([]string, error) {
+	o.Logger.Debugf("Listing routes")
+	result := []string{}
+	ctx, cancel := o.contextWithTimeout()
+	defer cancel()
+	req := o.computeSvc.Routes.List(o.ProjectID).Fields("items(name)").Filter(filter)
+	if len(filter) > 0 {
+		req = req.Filter(filter)
+	}
+	err := req.Pages(ctx, func(list *compute.RouteList) error {
+		for _, route := range list.Items {
+			o.Logger.Debugf("Found route: %s", route.Name)
+			result = append(result, route.Name)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to list routes")
+	}
+	return result, nil
+}
+
+func (o *ClusterUninstaller) deleteRoute(name string) error {
+	o.Logger.Debugf("Deleting route %s", name)
+	ctx, cancel := o.contextWithTimeout()
+	defer cancel()
+	_, err := o.computeSvc.Routes.Delete(o.ProjectID, name).Context(ctx).Do()
+	if err != nil && !isNoOp(err) {
+		return errors.Wrapf(err, "failed to delete route %s", name)
+	}
+	return nil
+}
+
+func (o *ClusterUninstaller) destroyRoutes() error {
+	routes, err := o.listClusterRoutes()
+	if err != nil {
+		return err
+	}
+	for _, route := range routes {
+		err = o.deleteRoute(route)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// getInternalLBInstanceGroups finds instance groups created for kube cloud controller
+// internal load balancers. They should be named "k8s-ig--{clusterid}":
+// https://github.com/openshift/kubernetes/blob/1e5983903742f64bca36a464582178c940353e9a/pkg/cloudprovider/providers/gce/gce_loadbalancer_naming.go#L33-L40
+// where clusterid is an 8-char id generated and saved in a configmap named
+// "ingress-uid" in kube-system, key "uid":
+// https://github.com/openshift/kubernetes/blob/1e5983903742f64bca36a464582178c940353e9a/pkg/cloudprovider/providers/gce/gce_clusterid.go#L210-L238
+// TODO: Have the installer generate this UID, save it to metadata and populate it in the cluster. With that, we can just directly get
+// an instanceGroup named "k8s-ig--{clusterid}"
+// For now, we look into each instance group and determine if it contains instances prefixed with the cluster's infra ID
+func (o *ClusterUninstaller) getInternalLBInstanceGroups() ([]nameAndZone, error) {
+	candidates, err := o.listInstanceGroups("name eq \"k8s-ig--.*\"")
+	if err != nil {
+		return nil, err
+	}
+	if len(candidates) == 0 {
+		return nil, nil
+	}
+
+	igName := ""
```

Comment:
why are we restricting to single instance group match?

Reply:
for internal load balancers created by the kube cloud controller, only a single instance group per zone is created per cluster. The same name is used for every zone. All we need to find is the one name and then find all instance groups named the same name in different zones.